### PR TITLE
yaml: detect and log warning on duplicate keys in snapcraft.yaml (CRAFT-553)

### PR DESCRIPTION
**Pull request by cjp256**
A very primitive check to warn if duplicate keys are found in
the snapcraft.yaml.  At the moment the file is loaded multiple
times which may result in duplicate warnings (ha ha) which we
need to address in future work.  Given that there shouldn't be
duplicates in YAML files, the extra verbosity should be OK.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

**Comment by cjp256** [comment]
Re-opened from #3517 to be on origin for edge publishing.

File: snapcraft/project/_project_info.py [hunk 1]
```diff
@@ -27,7 +27,9 @@ class ProjectInfo:
 
     def __init__(self, *, snapcraft_yaml_file_path) -> None:
         self.snapcraft_yaml_file_path = snapcraft_yaml_file_path
-        self.__raw_snapcraft = yaml_utils.load_yaml_file(snapcraft_yaml_file_path)
+        self.__raw_snapcraft = yaml_utils.load_yaml_file(
+            snapcraft_yaml_file_path, warn_duplicate_keys=True
```
Comment (cmatsuoka) [inline snapcraft/project/_project_info.py:31]: Is there a situation where we allow duplicate keys?

File: tests/unit/yaml_utils/test_yaml_utils.py [hunk 2]
```diff
@@ -15,39 +15,48 @@
...
-class OctIntTest(unit.TestCase):
-    def test_octint_dump(self):
-        output = io.StringIO()
-        yaml_utils.dump(dict(number=yaml_utils.OctInt(8)), stream=output)
-        output.seek(0)
+    assert [r.message for r in caplog.records] == []
 
-        self.assertThat(output.read().strip(), Equals("number: 0010"))
+
+def test_load_yaml_file_with_duplicate_warnings(caplog, tmp_path):
```
Comment (sergiusens) [inline tests/unit/yaml_utils/test_yaml_utils.py:45]: can you add a nest with nest duplicates?

File: tests/unit/yaml_utils/test_yaml_utils.py [hunk 3]
```diff
@@ -15,39 +15,48 @@
...
+    yaml_path = tmp_path / "test.yaml"
+    yaml_path.write_text("foo: bar\nfoo: bar2")
+
+    cfg = yaml_utils.load_yaml_file(str(yaml_path), warn_duplicate_keys=True)
```
Comment (cmatsuoka) [inline tests/unit/yaml_utils/test_yaml_utils.py:49]: If we make this configurable, we should also test for the situation where it's set to False.

File: snapcraft/project/_project_info.py (hunk 1 above)
Comment (cmatsuoka) [reply snapcraft/project/_project_info.py:31]: Removing this argument, warnings will be always enabled.

File: tests/unit/yaml_utils/test_yaml_utils.py (hunk 2 above)
Comment (cmatsuoka) [reply tests/unit/yaml_utils/test_yaml_utils.py:45]: Adding a test with the following scenario, is this what you mean? (In this case outer `foo` must not clash with inner `foo`, and inner `baz` must be detected)
```
foo: bar
nest:
  foo: bar2
  baz: bar3
  baz: bar4
```

File: tests/unit/yaml_utils/test_yaml_utils.py (hunk 3 above)
Comment (cmatsuoka) [reply tests/unit/yaml_utils/test_yaml_utils.py:49]: Removed configuration and leave it always enabled.

**Comment by cmatsuoka** [comment]
Rebasing on master because this branch is very old, merging master would make the commits hard to follow.

**Comment by cmatsuoka** [comment]
Original implementation only warned about the first duplicated key. Changed to issue warnings for all duplicated keys in the yaml file.

**Comment by cjp256** [comment]
👻
